#### Objective:
The categorical variables have now been replaced with their exact names.

It's observed that the features like Stalk color, Spore print color, Ring type and Population type may have a direct impact on the edibility of Mushrooms.

It's now required that we encode the variables and drop any unnecessary fields.

## Importing libraries and reading input CSV

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
from pandas_profiling import ProfileReport 
from sklearn.model_selection import train_test_split


Bad key savefig.frameon in file C:\Users\B.K.M.Bhat\Anaconda\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 421 ('savefig.frameon : True')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.3.1/matplotlibrc.template
or from the matplotlib source distribution

Bad key verbose.level in file C:\Users\B.K.M.Bhat\Anaconda\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 472 ('verbose.level  : silent      # one of silent, helpful, debug, debug-annoying')
You probably need to get an updated matplotlibrc file from
https://github.com/matplotlib/matplotlib/blob/v3.3.1/matplotlibrc.template
or from the matplotlib source distribution

Bad key verbose.fileo in file C:\Users\B.K.M.Bhat\Anaconda\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle, line 473 ('verbose.fileo  : sys.stdout  # a log filename, sys.stdout or sys.stderr')
You probably need to get an updated matp

In [5]:
mushroom_data = pd.read_csv('Mushrooms_cleansed.csv')
mushroom_data.head()

Unnamed: 0      class cap-shape cap-surface cap-color  bruises     odor  \
0           0  poisonous    convex      smooth     brown  bruises  pungent   
1           1     edible    convex      smooth    yellow  bruises   almond   
2           2     edible      bell      smooth     white  bruises    anise   
3           3  poisonous    convex       scaly     white  bruises  pungent   
4           4     edible    convex      smooth      gray       no     none   

  gill-attachment gill-spacing gill-size  ... stalk-surface-below-ring  \
0            free        close    narrow  ...                   smooth   
1            free        close     broad  ...                   smooth   
2            free        close     broad  ...                   smooth   
3            free        close    narrow  ...                   smooth   
4            free      crowded     broad  ...                   smooth   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                  white                  white   partial      white   
1                  white                  white   partial      white   
2                  white                  white   partial      white   
3                  white                  white   partial      white   
4                  white                  white   partial      white   

  ring-number   ring-type spore-print-color population  habitat  
0         one     pendant             black  scattered    urban  
1         one     pendant             brown   numerous  grasses  
2         one     pendant             brown   numerous  meadows  
3         one     pendant             black  scattered    urban  
4         one  evanescent             brown   abundant  grasses  

[5 rows x 24 columns]

### Checking for unique features

In [10]:
mushroom_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 24 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Unnamed: 0                8124 non-null   int64 
 1   class                     8124 non-null   object
 2   cap-shape                 8124 non-null   object
 3   cap-surface               8124 non-null   object
 4   cap-color                 8124 non-null   object
 5   bruises                   8124 non-null   object
 6   odor                      8124 non-null   object
 7   gill-attachment           8124 non-null   object
 8   gill-spacing              8124 non-null   object
 9   gill-size                 8124 non-null   object
 10  gill-color                8124 non-null   object
 11  stalk-shape               8124 non-null   object
 12  stalk-root                8124 non-null   object
 13  stalk-surface-above-ring  8124 non-null   object
 14  stalk-surface-below-ring

In [18]:
mushroom_data.nunique()

Unnamed: 0                  8124
class                          2
cap-shape                      6
cap-surface                    4
cap-color                     10
bruises                        2
odor                           9
gill-attachment                2
gill-spacing                   2
gill-size                      2
gill-color                    12
stalk-shape                    2
stalk-root                     5
stalk-surface-above-ring       4
stalk-surface-below-ring       4
stalk-color-above-ring         9
stalk-color-below-ring         9
veil-type                      1
veil-color                     4
ring-number                    3
ring-type                      5
spore-print-color              9
population                     6
habitat                        7
dtype: int64

All the variables are now categorical. As per the estimate, we are supposed to have 119 columns after encoding.

### One hot encoding all the variables

In [29]:
#One hot encoding done as all variables are categorical
mushroom_renamed_encoded = pd.get_dummies(mushroom_data)

In [20]:
mushroom_renamed_encoded.head()

Unnamed: 0  class_edible  class_poisonous  cap-shape_bell  \
0           0             0                1               0   
1           1             1                0               0   
2           2             1                0               1   
3           3             0                1               0   
4           4             1                0               0   

   cap-shape_conical  cap-shape_convex  cap-shape_flat  cap-shape_knobbed  \
0                  0                 1               0                  0   
1                  0                 1               0                  0   
2                  0                 0               0                  0   
3                  0                 1               0                  0   
4                  0                 1               0                  0   

   cap-shape_sunken  cap-surface_fibrous  ...  population_scattered  \
0                 0                    0  ...                     1   
1                 0                    0  ...                     0   
2                 0                    0  ...                     0   
3                 0                    0  ...                     1   
4                 0                    0  ...                     0   

   population_several  population_solitary  habitat_grasses  habitat_leaves  \
0                   0                    0                0               0   
1                   0                    0                1               0   
2                   0                    0                0               0   
3                   0                    0                0               0   
4                   0                    0                1               0   

   habitat_meadows  habitat_paths  habitat_urban  habitat_waste  habitat_woods  
0                0              0              1              0              0  
1                0              0              0              0              0  
2                1              0              0              0              0  
3                0              0              1              0              0  
4                0              0              0              0              0  

[5 rows x 120 columns]

There is an additional column Unnnamed: 0 added that needs to be dropped.

Also classes edible and poisonous are repetitive as a non-poisonous mushroom is always edible and vice versa. 
The column class_edible can be dropped.

The stalk_root_missing column can also be dropped.

In [21]:
mushroom_renamed_encoded.shape

(8124, 120)

The data contains 120 columns instead of 119 columns expected. 
This is due to the additional indexing column added.

In [30]:
mushroom_renamed_encoded.drop(['Unnamed: 0', 'class_edible', 'stalk-root_missing'], axis =1, inplace =True)

In [31]:
mushroom_renamed_encoded.shape

(8124, 117)

In [32]:
mushroom_renamed_encoded.head()

class_poisonous  cap-shape_bell  cap-shape_conical  cap-shape_convex  \
0                1               0                  0                 1   
1                0               0                  0                 1   
2                0               1                  0                 0   
3                1               0                  0                 1   
4                0               0                  0                 1   

   cap-shape_flat  cap-shape_knobbed  cap-shape_sunken  cap-surface_fibrous  \
0               0                  0                 0                    0   
1               0                  0                 0                    0   
2               0                  0                 0                    0   
3               0                  0                 0                    0   
4               0                  0                 0                    0   

   cap-surface_grooves  cap-surface_scaly  ...  population_scattered  \
0                    0                  0  ...                     1   
1                    0                  0  ...                     0   
2                    0                  0  ...                     0   
3                    0                  1  ...                     1   
4                    0                  0  ...                     0   

   population_several  population_solitary  habitat_grasses  habitat_leaves  \
0                   0                    0                0               0   
1                   0                    0                1               0   
2                   0                    0                0               0   
3                   0                    0                0               0   
4                   0                    0                1               0   

   habitat_meadows  habitat_paths  habitat_urban  habitat_waste  habitat_woods  
0                0              0              1              0              0  
1                0              0              0              0              0  
2                1              0              0              0              0  
3                0              0              1              0              0  
4                0              0              0              0              0  

[5 rows x 117 columns]

The columns for additional indexing, stalk root missing and class_edible have now been dropped.

### Train Test Split

In [86]:
#doing the train test split
mushroom_train,mushroom_test = train_test_split(mushroom_renamed_encoded, test_size=0.3)

In [87]:
mushroom_train.shape

(5686, 117)

In [88]:
mushroom_test.shape

(2438, 117)

In [89]:
mushroom_test.to_csv('mushroom_test.csv', index=False)
mushroom_train.to_csv('mushroom_train.csv', index=False)